### PR TITLE
coll/base: add knomial tree algorithm for MPI_Bcast

### PR DESCRIPTION
**Pull request by mkurnosov**
Implements k-nomial tree algorithm for `MPI_Bcast`.

Signed-off-by: Mikhail Kurnosov <mkurnosov@gmail.com>

### CODE REVIEW

**Comment by kawashima-fj** [comment]
ok to test

**Comment by bosilca** [comment]
We have a generic bcast function. You define a tree in terms of predecessor and list of successors, and the generic function (ompi_coll_base_bcast_intra_generic) will do the rest.

**Comment by mkurnosov** [comment]
@bosilca ok, I will change implementation

**Comment by bosilca** [comment]
Thanks @mkurnosov. Looks good.

**Comment by mkurnosov** [comment]
@bosilca I re-implemented algorithm for `base_bcast_intra_generic`. But, now it has limitations for radix value. The reason is fixed length of the `tree_next[MAXTREEFANOUT]` array. 
```
typedef struct ompi_coll_tree_t {
    int32_t tree_root;
    int32_t tree_fanout;
    int32_t tree_bmtree;
    int32_t tree_prev;
    int32_t tree_next[MAXTREEFANOUT];
    int32_t tree_nextsize;
} ompi_coll_tree_t;
```
Maximal number of child nodes in a k-nomial tree is (`radix-1)\ceil(\log_{radix}(comm_size)) = O(radix * log(comm_size))`. In worst case the algorithm switches to `base_bcast_intra_binomial`.

**Comment by bosilca** [comment]
We can bypass this limitation by reshaping the ompi_coll_tree_t structure, and allocate it in a smarter way.

```
typedef struct ompi_coll_tree_t {
    int32_t tree_root;
    int32_t tree_fanout;
    int32_t tree_bmtree;
    int32_t tree_prev;
    int32_t tree_nextsize;
    int32_t tree_next[1];
} ompi_coll_tree_t;
```
Then we can allocate it with ```malloc(sizeof(ompi_coll_tree_t) + (n-1) * sizeof(int32_t))```.

**Comment by mkurnosov** [comment]
@bosilca I will reshape the `ompi_coll_tree_t` structure and fix implementation

**Comment by mkurnosov** [comment]
@bosilca I reshaped `ompi_coll_tree_t` and fixed memory allocation for trees.

File: ompi/mca/coll/base/coll_base_topo.h [hunk 1]
```diff
@@ -23,15 +23,19 @@
...
 
 typedef struct ompi_coll_tree_t {
     int32_t tree_root;
     int32_t tree_fanout;
     int32_t tree_bmtree;
     int32_t tree_prev;
-    int32_t tree_next[MAXTREEFANOUT];
     int32_t tree_nextsize;
+    int32_t tree_next[1];
 } ompi_coll_tree_t;
```
Comment (hjelmn) [inline ompi/mca/coll/base/coll_base_topo.h]: Should be ```int32_t tree_next[];``` per C99.

File: ompi/mca/coll/base/coll_base_topo.h [hunk 2]
```diff
@@ -23,15 +23,19 @@
 
 #define MAXTREEFANOUT 32
 
+#define MEMBSIZE(type, member) (sizeof(((type *)0)->member))
+#define COLL_TREE_SIZE(fanout) \
```
Comment (hjelmn) [inline ompi/mca/coll/base/coll_base_topo.h]: If you update the struct as recommended you will also need to remove the ``` - 1``` here.

**Comment by mkurnosov** [comment]
@hjelmn I have changed the tree_next to a flexible array member

**Pull request merged**